### PR TITLE
Get document URL in esm-amd-loader

### PR DESCRIPTION
**Pull request by LarsDenBakker**
Currently the esm-amd-loader assumes that when it finds a top level module script, it's always a script in the index.html. When using HTML Imports (and possibility in the future HTML modules as well) this is not always the case. 

Relative module import specified in HTML imports are resolved properly for es modules, but when transpiled to AMD they are always resolved relative to the index.html. 

This PR makes it so that relative imports are resolved relative to the document (index or HTML import) that includes the script. It also makes it so that the correct import.meta.url is returned for these AMD modules.

I didn't change the tests yet, I first wanted to see if this was the right direction to fix HTML import - es module interop. 

Fixes https://github.com/Polymer/tools/issues/391

### CODE REVIEW

File: packages/esm-amd-loader/src/esm-amd-loader.ts [hunk 1]
```diff
@@ -421,14 +422,15 @@ window.define = function(deps: string[], moduleBody?: OnExecutedCallback) {
...
+  // Case #2: We are a top-level script in the HTML document or a HTML import.
+  // Resolve the URL relative to the document url. We can discover this case
+  // by waiting a tick, and if we haven't already been defined by the "onload"
+  // handler from case #1, then this must be case #2.
+  const documentUrl = getDocumentUrl();
   setTimeout(() => {
     if (defined === false) {
       pendingDefine = undefined;
-      const url = baseUrl + '#' + topLevelScriptIdx++ as NormalizedUrl;
+      const url = documentUrl + '#' + topLevelScriptIdx++ as NormalizedUrl;
```
[the remark this answers is not in the document]
Comment (LarsDenBakker) [reply packages/esm-amd-loader/src/esm-amd-loader.ts]: Good point, I've made the change.

File: packages/esm-amd-loader/src/esm-amd-loader.ts [hunk 2]
```diff
@@ -545,4 +547,37 @@ function getBaseUrl(): NormalizedUrl {
...
+      // If there is no import for the current script, we are in the index.html. Take the base url.
+      return baseUrl;
+    }
+    // Take the import href and strip off the filename
+    return htmlImport.href.substring(0, htmlImport.href.lastIndexOf('/') + 1);
+  } else {
+    // On chrome's native implementation it's not possible to get a direct reference to the link element,
+    // create a new script and let the browser resolve the url.
+    const script = currentScript.ownerDocument.createElement('script');
+    script.src = './';
```
[the remark this answers is not in the document]
Comment (LarsDenBakker) [reply packages/esm-amd-loader/src/esm-amd-loader.ts]: I didn't think '' would resolve to anything, but it did. It's better because it resolves to the filename.

File: packages/esm-amd-loader/src/esm-amd-loader.ts [hunk 3]
```diff
@@ -545,4 +547,37 @@ function getBaseUrl(): NormalizedUrl {
...
+
+  if (window.HTMLImports) {
+    // When the HTMLImports polyfill is active, we can take the path from the link element
+    const htmlImport = window.HTMLImports.importForElement(currentScript);
+    if (!htmlImport) {
+      // If there is no import for the current script, we are in the index.html. Take the base url.
+      return baseUrl;
+    }
+    // Take the import href and strip off the filename
+    return htmlImport.href.substring(0, htmlImport.href.lastIndexOf('/') + 1);
```
[the remark this answers is not in the document]
Comment (LarsDenBakker) [reply packages/esm-amd-loader/src/esm-amd-loader.ts]: Turns out we can just return the href here, this is in line with what we're returning from the native implementation as well.

**Comment by LarsDenBakker** [comment]
@aomarks I've made the changes and added tests. 

I couldn't get the tests to run from a html import, and I didn't want to introduce too much deviation from the current set up just for html imports. Creating the imports imperatively does the job pretty simply as well. What do you think?

I added the v1 webcomponents loader, as the v2 loader does not add html imports.

**Comment by LarsDenBakker** [comment]
The tests run fine for me locally, but I'm not sure what's going wrong inside appveyor and travis?

File: packages/esm-amd-loader/tsconfig.json [hunk 4]
```diff
@@ -7,7 +7,7 @@
     "preserveConstEnums": false,
     "sourceMap": false,
     "lib": [
-      "es5",
+      "es2015",
```
[the remark this answers is not in the document]
Comment (LarsDenBakker) [reply packages/esm-amd-loader/tsconfig.json]: Sorry, I didn't mean to change that.

File: packages/esm-amd-loader/test/src/suite.ts [hunk 5]
```diff
@@ -393,3 +397,47 @@ suite('cyclical dependencies', () => {
...
+    testImport('../html-import/y/deep-import.html', ['x', 'z', 'y', 'deep-import'], done);
+  });
+
+  test('imports with child imports', (done) => {
+    testImport('../html-import/parent-import.html', ['z', 'y', 'child-import', 'x', 'parent-import'], done);
+  });
+
+  test('import with meta', (done) => {
+    window.testImportMeta = (url) => {
+        assert.match(url, /https?:\/\/.+\/html-import\/meta/);
```
[the remark this answers is not in the document]
Comment (LarsDenBakker) [reply packages/esm-amd-loader/test/src/suite.ts]: You're right, made the change.

File: packages/esm-amd-loader/test/src/suite.ts [hunk 6]
```diff
@@ -393,3 +397,47 @@ suite('cyclical dependencies', () => {
...
+      link.rel = 'import';
+      link.href = href;
+      document.head.appendChild(link);
+  }
+
+  function testImport(href: string, expectedOrder: string[], done: () => void) {
+      // Each time and amd module in the chain is executed, it registers itself.
+      // If we've reached the length of modules we are expecing to be loaded,
+      // we check if the right modules were loaded in the expected order
+      window.addExcutedForImport = (name: string) => {
```
[the remark this answers is not in the document]
Comment (LarsDenBakker) [reply packages/esm-amd-loader/test/src/suite.ts]: Done

File: packages/esm-amd-loader/test/src/suite.ts [hunk 7]
```diff
@@ -393,3 +397,47 @@ suite('cyclical dependencies', () => {
...
+suite('html imports', () => {
+  function importHref(href: string) {
+      const link = document.createElement('link');
+      link.rel = 'import';
+      link.href = href;
+      document.head.appendChild(link);
+  }
+
+  function testImport(href: string, expectedOrder: string[], done: () => void) {
+      // Each time and amd module in the chain is executed, it registers itself.
```
[the remark this answers is not in the document]
Comment (LarsDenBakker) [reply packages/esm-amd-loader/test/src/suite.ts]: Done

File: packages/esm-amd-loader/test/static/html-import/meta/import-meta.html [hunk 8]
```diff
@@ -0,0 +1,7 @@
+<link rel="import" href="./y/child-import.html">
+
+<script>
+  define(['meta'], (meta) => {
+    window.testImportMeta(meta.url);
```
[the remark this answers is not in the document]
Comment (LarsDenBakker) [reply packages/esm-amd-loader/test/static/html-import/meta/import-meta.html]: I think this is already the case. In the start of the test we make a global function which is called from the import. If the module file fails to load or if the code fails to run, the function isn't called and thus done isn't called. This causes the test to fail.

**Comment by LarsDenBakker** [comment]
@aomarks Made the changes. Also, I've run the tests on chrome, firefox, safari, edge and IE11.
